HLTriggerOffline/Scouting/plugins/PatElectronTagProbeAnalyzer.cc

```diff
@@ -0,0 +1,843 @@
+  return pt;
+}
+
+bool PatElectronTagProbeAnalyzer::scoutingElectronGsfTrackID(const Run3ScoutingElectron el, size_t trackIdx) const {
+  if (trackIdx > el.trkpt().size())
+    edm::LogError("ScoutingMonitoring") << "Invalid track index for electron: Exceeds the number of tracks";
+
+  math::PtEtaPhiMLorentzVector particleSC(el.pt(), el.eta(), el.phi(), 0.0005);
+  math::PtEtaPhiMLorentzVector particleTrk(el.trkpt()[trackIdx], el.trketa()[trackIdx], el.trkphi()[trackIdx], 0.0005);
+
+  double scEnergy = particleSC.energy();
+  double trkEnergy = particleTrk.energy();
+  double relEnergyDiff = fabs(scEnergy - trkEnergy) / scEnergy;
+  double dPhi = deltaPhi(particleSC.phi(), particleTrk.phi());
+
+  bool isEB = (fabs(el.eta()) < 1.479);
+  if (isEB) {
+    if (el.trkpt()[trackIdx] < 12)
+      return false;
+    if (relEnergyDiff > 1)
+      return false;
+    if (dPhi > 0.06)
+      return false;
+    if (el.trkchi2overndf()[trackIdx] > 3)
+      return false;
+    return true;
+  } else {
+    if (el.trkpt()[trackIdx] < 12)
+      return false;
+    if (relEnergyDiff > 1)
+      return false;
+    if (dPhi > 0.06)
+      return false;
+    if (el.trkchi2overndf()[trackIdx] > 2)
+      return false;
+    return true;
+  }
+}
+
+bool PatElectronTagProbeAnalyzer::scoutingElectronGsfTrackIdx(const Run3ScoutingElectron el, size_t& trackIdx) const {
+  bool foundGoodGsfTrkIdx = false;
+  for (size_t i = 0; i < el.trkpt().size(); ++i) {
+    if (scoutingElectronGsfTrackID(el, i)) {
+      if (!foundGoodGsfTrkIdx) {
+        foundGoodGsfTrkIdx = true;
+        trackIdx = i;
+      } else {
+        double relPtDiff = fabs(el.trkpt()[i] - el.pt()) / el.pt();
+        double relPtDiffOld = fabs(el.trkpt()[trackIdx] - el.pt()) / el.pt();
+        if (relPtDiff < relPtDiffOld)
+          trackIdx = i;
+      }
+    }
+  }
+  return foundGoodGsfTrkIdx;
+}
+
+bool PatElectronTagProbeAnalyzer::hasPatternInHLTPath(const edm::TriggerNames& triggerNames,
+                                                      const std::string& pattern) const {
+  for (unsigned int i = 0; i < triggerNames.size(); ++i) {
+    const std::string& triggerName = triggerNames.triggerName(i);
+
+    // Check if triggerName starts with the specified prefix
+    if (triggerName.find(pattern) == 0) {  // Position 0 means it starts with 'prefix'
+      return true;                         // Pattern match found
+    }
+  }
+  return false;  // No match found
+}
+
+bool PatElectronTagProbeAnalyzer::scoutingElectron_passHLT(float el_eta,
+                                                           float el_phi,
+                                                           TString filter,
+                                                           trigger::TriggerObjectCollection legObjects) const {
+  bool foundTheLeg = false;
+  for (unsigned int i = 0; i < legObjects.size(); i++) {
+    float delR = deltaR(legObjects.at(i).eta(), legObjects.at(i).phi(), el_eta, el_phi);
```

Comment:
the `deltaR2` interface is faster (saves a `sqrt`):

https://github.com/cms-sw/cmssw/blob/902cf5d1ca29a9d77a771245d2cfddf569f5eb59/DataFormats/Math/interface/deltaR.h#L16